### PR TITLE
ARTEMIS-3271 Improve Critical Analyzer to use AutoCloseable on the API

### PR DESCRIPTION
**Pull request by clebertsuconic**

### CODE REVIEW

File: artemis-commons/src/main/java/org/apache/activemq/artemis/utils/critical/CriticalMeasure.java [hunk 1]
```diff
@@ -28,71 +27,58 @@
...
 
-   //uses updaters to avoid creates many AtomicLong instances
-   static final AtomicLongFieldUpdater<CriticalMeasure> TIME_ENTER_UPDATER = AtomicLongFieldUpdater.newUpdater(CriticalMeasure.class, "timeEnter");
+   AutoCloseable autoCloseable = () -> {
+      leaveCritical();
+      CURRENT_AUTO_CLOSEABLE.set(this, null);
+   };
 
-   static final AtomicReferenceFieldUpdater<CriticalMeasure, Thread> CURRENT_THREAD_UDPATER = AtomicReferenceFieldUpdater.newUpdater(CriticalMeasure.class, Thread.class, "currentThread");
+   public static AutoCloseable dummyCloseable = () -> { };
```
Comment (franz1981) [inline artemis-commons/src/main/java/org/apache/activemq/artemis/utils/critical/CriticalMeasure.java:37]: Static final

**Comment by franz1981** [comment]
You can create a CriticalSection interface that extends AutoCloseable and remove the Exception from ita signature. It would simplify the caller, saving the catch path.

File: artemis-commons/src/main/java/org/apache/activemq/artemis/utils/critical/CriticalMeasure.java (hunk 1 above)
Comment (clebertsuconic) [reply artemis-commons/src/main/java/org/apache/activemq/artemis/utils/critical/CriticalMeasure.java:37]: oops.. thanks

**Comment by clebertsuconic** [comment]
> You can create a CriticalSection interface that extends AutoCloseable and remove the Exception from its signature, by overriding close. It would simplify the caller, saving the catch path.

nice idea! thanks

**Review by franz1981 — APPROVED** [review]
LGTM
It's a super nice improvement in term of readability and to save forgetting to exit :)

**Comment by franz1981** [comment]
@clebertsuconic 
I'm getting some weird issue while under heavy load...
```
2021-05-15 10:00:49,256 WARN  [org.apache.activemq.artemis.utils.critical.CriticalMeasure] Component org.apache.activemq.artemis.core.server.impl.QueueImpl is expired on path 0
2021-05-15 10:00:49,258 INFO  [org.apache.activemq.artemis.core.server] AMQ224107: The Critical Analyzer detected slow paths on the broker.  It is recommended that you enable trace logs on org.apache.activemq.artemis.utils.critical while you troubleshoot this issue. You should disable the trace logs when you have finished troubleshooting.
2021-05-15 10:00:49,258 ERROR [org.apache.activemq.artemis.core.server] AMQ224079: The process for the virtual machine will be killed, as component QueueImpl[name=q_7, postOffice=PostOfficeImpl [server=ActiveMQServerImpl::serverUUID=64e3f254-b585-11eb-8f02-000f532921b8], temp=false]@9a33f70 is not responsive
2021-05-15 10:00:50,429 WARN  [org.apache.activemq.artemis.core.server] AMQ222199: Thread dump: *******************************************************************************
Complete Thread dump 
```
```
"Thread-17 (ActiveMQ-server-org.apache.activemq.artemis.core.server.impl.ActiveMQServerImpl$6@15bbf42f)" Id=142 RUNNABLE
	at io.netty.channel.epoll.Native.eventFdWrite(Native Method)
	at io.netty.channel.epoll.EpollEventLoop.wakeup(EpollEventLoop.java:186)
	at io.netty.util.concurrent.SingleThreadEventExecutor.execute(SingleThreadEventExecutor.java:849)
	at io.netty.util.concurrent.SingleThreadEventExecutor.execute(SingleThreadEventExecutor.java:818)
	at io.netty.channel.AbstractChannelHandlerContext.safeExecute(AbstractChannelHandlerContext.java:989)
	at io.netty.channel.AbstractChannelHandlerContext.write(AbstractChannelHandlerContext.java:796)
	at io.netty.channel.AbstractChannelHandlerContext.writeAndFlush(AbstractChannelHandlerContext.java:758)
	at io.netty.channel.DefaultChannelPipeline.writeAndFlush(DefaultChannelPipeline.java:1020)
	at io.netty.channel.AbstractChannel.writeAndFlush(AbstractChannel.java:311)
	at org.apache.activemq.artemis.core.remoting.impl.netty.NettyConnection.write(NettyConnection.java:364)
	at org.apache.activemq.artemis.core.remoting.impl.netty.NettyConnection.write(NettyConnection.java:288)
	at org.apache.activemq.artemis.core.protocol.core.impl.ChannelImpl.send(ChannelImpl.java:418)
	-  locked java.lang.Object@7755ba9d
	at org.apache.activemq.artemis.core.protocol.core.impl.ChannelImpl.sendBatched(ChannelImpl.java:329)
	at org.apache.activemq.artemis.core.protocol.core.impl.CoreSessionCallback.sendMessage(CoreSessionCallback.java:128)
	at org.apache.activemq.artemis.core.server.impl.ServerConsumerImpl.deliverStandardMessage(ServerConsumerImpl.java:1170)
	at org.apache.activemq.artemis.core.server.impl.ServerConsumerImpl.proceedDeliver(ServerConsumerImpl.java:514)
	at org.apache.activemq.artemis.core.server.impl.QueueImpl.proceedDeliver(QueueImpl.java:3747)
	at org.apache.activemq.artemis.core.server.impl.QueueImpl.deliver(QueueImpl.java:3689)
	-  locked org.apache.activemq.artemis.core.server.impl.QueueImpl@9a33f70
	at org.apache.activemq.artemis.core.server.impl.QueueImpl.deliverDirect(QueueImpl.java:3643)
	at org.apache.activemq.artemis.core.server.impl.QueueImpl.addTail(QueueImpl.java:1218)
	at org.apache.activemq.artemis.core.server.impl.RoutingContextImpl.internalprocessReferences(RoutingContextImpl.java:182)
	at org.apache.activemq.artemis.core.server.impl.RoutingContextImpl.processReferences(RoutingContextImpl.java:177)
	at org.apache.activemq.artemis.core.postoffice.impl.PostOfficeImpl$2.done(PostOfficeImpl.java:1568)
	at org.apache.activemq.artemis.core.persistence.impl.journal.OperationContextImpl$1.run(OperationContextImpl.java:279)
	at org.apache.activemq.artemis.utils.actors.OrderedExecutor.doTask(OrderedExecutor.java:42)
	at org.apache.activemq.artemis.utils.actors.OrderedExecutor.doTask(OrderedExecutor.java:31)
	at org.apache.activemq.artemis.utils.actors.ProcessorBase.executePendingTasks(ProcessorBase.java:65)
	at org.apache.activemq.artemis.utils.actors.ProcessorBase$$Lambda$37/599984672.run(Unknown Source)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at org.apache.activemq.artemis.utils.ActiveMQThreadFactory$1.run(ActiveMQThreadFactory.java:118)

	Number of locked synchronizers = 2
	- java.util.concurrent.locks.ReentrantLock$NonfairSync@2e770b23
	- java.util.concurrent.ThreadPoolExecutor$Worker@6cf16158
```
The stack trace show just a single thread in `RUNNABLE` state in that same path...so probably there is some error related how we collect the metric...

**Comment by clebertsuconic** [comment]
Seems a legit lock issue to me.

**Comment by franz1981** [comment]
@clebertsuconic it appear only when I push some load with many clients and older broker versions seems not affected by this...do we have recently added any lock in the expired path?

**Comment by clebertsuconic** [comment]
@franz1981 try an older Netty version?

**Comment by clebertsuconic** [comment]
@franz1981 ChannelImpl is an area of changes... they were meant for replication but it seems it's affecting it here.

**Comment by franz1981** [comment]
@clebertsuconic good point, let me open a separate issue with this and if can be replicated without replication.
The channel impl send batched path up shouldn't be changed and Netty eventFdWrite is a JNI call that cannot block AFAIK. I am going to attach any other stack trace I get while running this so we can fix it

**Comment by clebertsuconic** [comment]
@franz1981 disable direct deliver on the broker.

**Comment by clebertsuconic** [comment]
Perhaps you found an issue on critical analyzer.  Dunno.  Worth to check it.

**Comment by clebertsuconic** [comment]
What version was this ?


Try disabling direct deliver.  When doing extreme large horizontal loads with many consumers   is expected to have to disable that.

**Comment by franz1981** [comment]
It's latest main using 16 producers/consumers sending/consuming 100 byte messages for long time (more then 5 minutes) on 16 different queues (auto-created).
I am going to try it again today and will open an issue with my findings..
The difference is that I am using separate big machines for everything: master, slave and clients

**Pull request merged**